libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,149 @@
+
+using namespace std;
+
+namespace solidity::frontend
+{
+
+
+shared_ptr<FunctionCallGraphBuilder::ContractCallGraph> const FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	m_contract = &_contract;
+
+	std::shared_ptr<ContractCallGraph> graph= make_shared<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	m_nodes = &graph->creationCalls;
+	visitConstructor(_contract);
+
+
+	// Create graph for all publicly reachable functions
+	m_nodes = &graph->runtimeCalls;
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!m_nodes->count(funcDef))
+				visitCallable(funcDef);
+
+	m_contract = nullptr;
+	m_nodes = nullptr;
+	solAssert(m_currentNode == nullptr, "Current node not properly reset.");
+
+	return graph;
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		if (*_identifier.annotation().requiredLookup == VirtualLookup::Virtual)
+			callable = &callable->resolveVirtual(*m_contract);
+
+		solAssert(
+			*_identifier.annotation().requiredLookup != VirtualLookup::Super,
+			"Only MemberAccess can have lookup 'super'"
+		);
+
+		// Ignore calls to other contract
+		if (callable->annotation().contract && !m_contract->derivesFrom(*callable->annotation().contract) && !callable->annotation().contract->isLibrary())
+			return true;
+
+		if (!m_nodes->count(callable))
+			visitCallable(callable);
+
+		m_currentNode->calls.emplace(*m_nodes->find(callable));
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+	{
+		CallableDeclaration const* callable = contractType->contractDefinition().constructor();
+
+		if (!m_nodes->count(callable))
+			m_nodes->emplace(make_shared<Node>(callable));
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(MemberAccess const& _memberAccess)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_memberAccess.annotation().referencedDeclaration))
+	{
+		if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+		{
+			ContractDefinition const* super =
+				m_currentNode->callable ?
+				m_currentNode->callable->annotation().contract :
+				m_contract;
+
+			callable = &callable->resolveVirtual(*m_contract, super);
+		}
+
+		solAssert(
+			*_memberAccess.annotation().requiredLookup != VirtualLookup::Virtual,
+			"MemberAccess cannot have lookup 'virtual'"
+		);
+
+		// Ignore calls to other contracts
+		if (callable->annotation().contract && !m_contract->derivesFrom(*callable->annotation().contract) && !callable->annotation().contract->isLibrary())
```

Comment:
This is a bit unfortunate. What we want to know is if this is an internal or an external function call. Can you check if the type of the expression is an external or an internal function type?

Reply:
and ignore if it is an external type?

Reply:
Yes, ignore external calls.